csrc/kernel_cache.cpp

```diff
@@ -779,10 +779,31 @@ FusionKernelRuntime* FusionExecutorCache::getKernelRuntimeFor(
   // that whenever we encounter a new set of input shapes we segment and compile
   // a new FusionKernelRuntime.
   if (!isOptionDisabled(DisableOption::KernelReuse)) {
+    std::optional<bool> should_not_be_segmented = std::nullopt;
     auto reuse_it = std::find_if(
         kernel_runtimes.begin(),
         kernel_runtimes.end(),
-        [&args, &new_heuristics, &forced_index_type](auto& kernel_runtime) {
+        [&should_not_be_segmented, &args, &new_heuristics, &forced_index_type](
+            auto& kernel_runtime) {
+          if (!should_not_be_segmented.has_value()) {
+            // Use the concretized complete fusion from this runtime to check
+            // whether we can schedule the complete fusion without segmenting.
+            std::unique_ptr<Fusion> conc_fusion = std::make_unique<Fusion>(
+                *(kernel_runtime->fusionSegments()->completeFusion()));
+            SchedulerRuntimeInfo runtime_info(
+                conc_fusion.get(),
+                args,
+                /*precomputed_values=*/nullptr,
+                /*all_tvs=*/{},
+                forced_index_type);
+            should_not_be_segmented = SchedulerEntry::proposeHeuristics(
+                                          conc_fusion.get(), runtime_info)
+                                          .has_value();
+          }
+          if (should_not_be_segmented.value() &&
+              kernel_runtime->isSegmented()) {
+            return false;
+          }
           auto maybe_heuristics =
```

Comment:
After found a heuristic for this complete fusion, can we directly use it?

Reply:
Update: this is actually not getting the heuristic for the unsegmented fusion. It is merely running `canSchedule` on the fusion and returning the first `ScheduleHeuristic` (enum) that succeeds. Later we run `kernel_runtime->getMaybeHeuristicsFor` which will run `SegmentedGroup::getMaybeSchedulerEntry` to actually get a config. It's possible to refactor this to reuse that call so that we don't need to re-run the heuristic on unsegmented runtimes, but I think it would duplicate more of the segmentation machinery.